libherokubuildpack/src/tgz.rs

```diff
@@ -0,0 +1,233 @@
+use flate2::read::GzDecoder;
+use sha2::{
+    digest::{generic_array::GenericArray, OutputSizeUser},
+    Digest, Sha256,
+};
+use std::path::PathBuf;
+use std::{io::Read, path::StripPrefixError};
+use tar::Archive;
+
+/// Fetches a tarball (.tgz or .tar.gz) from a `remote_uri` and extracts the
+/// contents to a `dest_dir`. Care is taken not to write temporary files or read
+/// the entire contents into memory. It will also optionally `strip_prefix`
+/// similar to `tar`'s `--strip-components` flag, `filter_prefixes` to prevent
+/// superflous files from being written, and `verify_digest` against a known
+/// SHA256 digest.
+#[derive(Debug, Clone, Default)]
+pub struct Fetcher {
+    pub remote_uri: String,
+    pub dest_dir: std::path::PathBuf,
+    pub strip_prefix: Option<String>,
+    pub filter_prefixes: Option<Vec<String>>,
+    pub verify_digest: Option<String>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("HTTP error while fetching archive: {0}")]
+    Http(#[from] Box<ureq::Error>),
+    #[error("Error reading archive entries: {0}")]
+    Entries(std::io::Error),
+    #[error("Error reading archive entry: {0}")]
+    Entry(std::io::Error),
+    #[error("Error reading archive file path: {0}")]
+    Path(std::io::Error),
+    #[error("Failed to validate archive checksum; expected {0}, but found {1}")]
+    Checksum(String, String),
+    #[error("Error writing archive entry: {0}")]
+    Unpack(std::io::Error),
+    #[error("Error stripping archive entry prefix: {0}")]
+    Prefix(StripPrefixError),
+}
+
+impl Fetcher {
+    /// Initialize a Fetcher with the minimum arguments.
+    #[must_use]
+    pub fn new<S: Into<String>, P: Into<PathBuf>>(remote_uri: S, dest_dir: P) -> Self {
+        Self {
+            remote_uri: remote_uri.into(),
+            dest_dir: dest_dir.into(),
+            strip_prefix: Option::default(),
+            filter_prefixes: Option::default(),
+            verify_digest: Option::default(),
+        }
+    }
+
+    /// Intruct the fetcher to verify that the tarball digest matches a known SHA256 digest.
+    #[must_use]
+    pub fn verify_digest<S: Into<String>>(&mut self, digest: S) -> &mut Self {
+        self.verify_digest = Some(digest.into());
+        self
+    }
+
+    /// Instruct the Fetcher to only extract files matching prefixes.
+    #[must_use]
+    pub fn filter_prefixes<I: IntoIterator<Item = S>, S: Into<String>>(
```

Comment:
instead of just allowing for a prefix check, would it not be more flexible to allow for a closure to be passed where you could perform any action you want to check if a file should be included or not?  e.g.;

```rust
Fetcher::new(
    "https://mirrors.edge.kernel.org/pub/software/scm/git/git-0.01.tar.gz",
    dest.clone(),
)
.strip_prefix("git-0.01")
.filter(|path| path.starts_with("README"))
.verify_digest("9bdf8a4198b269c5cbe4263b1f581aae885170a6cb93339a2033cb468e57dcd3")
.fetch()
```